### PR TITLE
fix(projector): preserve authoritative B2AGG LET indices across restart and restore

### PR DESCRIPTION
**Pull request by mandrigin**
## Summary

Prevents the synthetic projector from assigning or exposing an incorrect B2AGG `deposit_count` / `globalIndex` when several exits are consumed in one transaction, after a restart, or during restore.

The projector now proves the complete execution order and Local Exit Tree (LET) accounting before it exposes a synthetic block. If identity, body, ordering, or cardinality evidence is incomplete, it fails the tick before sealing.

```mermaid
flowchart LR
    NS[Miden note sweep] --> ID[Append-only nullifier to NoteId ledger]
    TX[Bridge TransactionRecord feed] --> XO[Order transactions by account-state chain]
    XO --> IO[Enumerate transaction input_notes in execution order]
    ID --> BODY[Fetch canonical public note body]
    IO --> BODY
    BODY --> RES[Reserve exact LET index by unique NoteId]
    RES --> NG{Per-note emission gates}
    NG -->|eligible| EVT[Atomic BridgeEvent commit]
    NG -->|quarantine / defer / self-target| NOEVT[Keep reservation, emit no event]
    EVT --> CG{On-chain LET leaves equal baseline plus reservations?}
    NOEVT --> CG
    CG -->|yes| SEAL[Expose complete synthetic block]
    CG -->|no| STOP[Fail tick; expose nothing]
```

### Authoritative execution order

- Reconstructs same-block bridge transaction order from each account update's initial/final state commitments instead of RPC response or transaction-hash order.
- Uses `TransactionRecord.transaction_header.input_notes()` sequence for within-transaction order. This is the transaction-committed consumption order; it is not a locally invented sort.
- Uses unique `NoteId` for reservation, deduplication, quarantine, and ordering identity. Details commitments remain only a deterministic fallback tie-breaker and preserve the historical synthetic transaction-hash scheme.
- Fails closed when same-transaction siblings cannot be ordered.

### Restart-safe identity and body recovery

- Persists the minimal B2AGG `nullifier -> NoteId` identity ledger because pinned `miden-client` strips input headers after consumption and its local details-keyed representation can collapse same-details siblings.
- Keeps that ledger append-only; canonical note bodies continue to come from the Miden node.
- Requires exact aggregate `get_notes_by_id` responses before advancing the reconciliation cursor.
- Treats an identified body omission as a hard tick failure, including when an earlier attempt already reserved the leaf.
- Holds projection behind a full-tip note-visibility barrier.

### Exact LET reservations and pre-seal gate

- Reserves an absolute LET index for every bridge-consumed B2AGG before metadata, quarantine, self-target, or other non-emission gates.
- Separates `reserved` from `emitted`; a skipped leaf retains its real position and the next valid event receives the next real index.
- Validates existing reservations are a contiguous execution-order prefix with the expected absolute indices.
- Seals only when:

```text
on-chain LET leaves == audited legacy baseline + durable reservations
```

### Restore and upgrade behavior

- Restore reads the authoritative bridge transaction/body stream directly, preserves full `(block, transaction order, input position, details commitment, NoteId)` order, requires exact body/cardinality evidence, and shares the live B2AGG derivation.
- Full `--restore` requires `let_gate_baseline = 0`; a nonzero legacy offset needs offline history reconstruction first.
- Migration `016` adds reservation state and the explicit audited baseline.
- Migration `017` intentionally resets `reconcile_cursor` once so an upgraded deployment performs a full historical identity backfill. Projection stays held until that sweep reaches the current tip; subsequent starts resume normally.

## Validation

- `make lint` — passed
- `make test-unit` — 434 passed, 1 intentionally ignored
- Focused projector tests — 33 passed
- Focused restore tests — 35 passed
- `cargo check --all-targets --features postgres` — passed
- E2E was not run in this patch cycle; the release/e2e agent can run it after merge readiness.

## Scope

The projector remains single-instance. Coordinating multiple proxy replicas against one PostgreSQL store is tracked separately and is intentionally out of scope for this PR.

### CODE REVIEW

**Review by mandrigin — COMMENTED** [review]
Blocking early review of `7f72c0a`. The architecture is directionally right, but the claimed fail-closed safety property is not established:

1. **A prior divergence is cleared on an off-tip tick.** `LetGateState::evaluate` returns `Proceed` whenever `!at_tip`; the caller clears `let_gate_blocked` and then seals through `project_to`. An at-tip invisible gap can therefore block one tick and be permanently sealed on the next tick if the visibility barrier lags while the chain advances, without any aligned/healed verdict.

2. **Production restore still uses hash order.** `sort_consumed_for_projection` hard-wires an empty position map, and the real restore path calls that wrapper. The parity test injects a fabricated non-empty map directly, so it never exercises production restore.

3. **The documented/default launch disables the gate.** With no explicit `--miden-node`, `MidenClient` correctly defaults to localhost, but `SyntheticProjector::new` maps the absent option to `node_rpc=None`; the new gate only runs when that option is `Some`.

4. **Restart clears a halt.** Gate state is memory-only. First at-tip arming absorbs every standing `on_chain - visible` gap into the baseline and proceeds, so restarting after a halt—or discovering an omission during initial catch-up—turns the unsafe gap into accepted history.

5. **Within-tx positions use non-unique `NoteDetailsCommitment`.** Distinct notes with different metadata can share details. Map insertion overwrites one position, both records appear “resolved,” and sorting becomes incorrect. Projection also deduplicates on this commitment while the cardinality gate counts both, so the gate can report aligned while only one of two LET leaves is emitted. Use a unique NoteId/nullifier identity end to end.

6. **Quarantine accounting can double-count.** The monitor adds an unfiltered global count of all quarantine rows. Restore recovery has no delete/supersede store API, so a recovered exit can count once as emitted and again as quarantined; rows belonging to an older bridge are also included.

The seven new unit tests cover direct helpers but not these integrations: no tick-level blocked→off-tip case, restart-after-halt, default config, actual restore position path, unique-identity collision, PostgreSQL quarantine accounting, or multi-input E2E.

Please keep this out of merge until those paths are fixed. CI is green, but it does not exercise the unsafe state transitions.

**Review by mandrigin — COMMENTED** [review]
Additional blocking correctness issue on current head `7f72c0a`:

**Skipped LET leaves do not reserve their `deposit_count` / LET index.** The pre-seal gate deliberately counts every feed-visible B2AGG consumption, including quarantined/deferred/self-targeted exits, and `commit_sealed` folds that count into its cumulative state. But the actual allocation counter advances only inside `commit_b2agg_event_atomic`, i.e. only when a `BridgeEvent` is emitted.

Concrete failure: LET leaf 0 is quarantined, then leaf 1 is valid. The first exit emits nothing and does not advance `deposit_counter`; the valid exit is then emitted with `deposit_count = 0` although its authoritative LET index is 1. That produces the wrong `globalIndex`. The post-emit monitor can hide this because it compares `deposit_counter + quarantine_count` to the on-chain leaf count.

The quarantine test currently stops after asserting that the quarantined exit emitted nothing; it never emits the next valid leaf and checks its index.

Please reserve/persist the authoritative LET position for **every** consumed B2AGG leaf, including non-emittable classes, or otherwise allocate valid events from authoritative LET ordering rather than an emitted-event-only counter. Add a regression with `quarantine -> valid` (and restart/retry) asserting the valid event has `deposit_count = 1` and the correct `globalIndex`.

**Review by mandrigin — COMMENTED** [review]
Re-reviewed through `47566bd`. Several prior blockers are genuinely fixed: off-tip ticks no longer clear standing divergence; clean-state skipped leaves reserve indices; quarantine is no longer double-counted; the live path uses unique `NoteId`; and the projector's default node URL now arms the gate. The latest implementation is still not safe to merge, however:

1. **Migration 011 is never run.** `migrations/011_deposit_reservations.sql` adds `emitted`, `let_gate_baseline`, and `let_gate_halted`, but `src/store/migrator.rs::MIGRATIONS` still ends at `010_reconcile_cursor.sql`. A normal PostgreSQL startup therefore reaches code that selects/updates columns the migrator never created. Add the migration to the embedded ordered list (and coordinate/renumber it against #123's independent `011_claim_mint_serials.sql`).

2. **A positive baseline still produces the wrong next `deposit_count`.** The gate accepts `on_chain = baseline + deposit_counter + pending`, but both `reserve_deposit_index` implementations return raw `deposit_counter`. With baseline `S > 0`, the next true LET index is `S + deposit_counter`, while the emitted event receives `deposit_counter`. The state-machine test arms at baseline 97 but never projects a note or asserts its index. Make index allocation consistent with the persisted baseline and add a regression that arms with a non-zero baseline, projects the next leaf, and checks its exact `deposit_count`/global index.

3. **Default restore still bypasses the effective node URL.** `main` passes `command.miden_node.as_deref()` to `restore`, so the documented default launch supplies `None` and builds no NoteId/position map. A live NoteId-keyed reservation replayed with `unique_id=None` falls back to the commitment key, misses the existing row, and can reserve/emit again. Pass the same effective URL used by the live projector and fail closed when authoritative identity cannot be recovered.

4. **Restore discards authoritative block/transaction order.** `restore_bridge_positions` returns `(commitment, block, tx_order, within_tx_pos)`, but `restore_bridge_outs` retains only `within_tx_pos`; the comparator takes block/order from the local consumed record. Imported/historical records can lack those fields, causing notes from multiple transactions in one block to fall through to position/hash order. Carry the full authoritative tuple into the shared sort key. The current parity test fabricates populated local order fields and does not cover this production shape.

5. **Restore does not actually fail closed on an incomplete position scan.** Missing identities become `None`, position defaults to zero, and restore proceeds to sort/project. The ambiguity guard exists only in `SyntheticProjector::project_block_notes`, which restore does not call. An incomplete or failed position feed must stop B2AGG replay before reserving or emitting anything; add partial-feed and same-tx ambiguity regressions through the real restore path.

6. **A transient baseline-persistence failure can be forgotten.** `LetGateState::evaluate` mutates the in-memory baseline before `set_let_gate_baseline().await`. If that store write fails, later ticks see an already-armed gate and do not retry persistence; after restart the service can re-arm and absorb a gap. Persist before committing the in-memory arming transition, or retain explicit “persistence pending” state and retry.

The branch's unit/security checks are green and GitHub currently reports it mergeable, but the L2↔L2 E2E is skipped and these are runtime/index-integrity blockers. There are no unresolved inline threads; this review supersedes the earlier top-level blocker list for the current head.

**Review by mandrigin — COMMENTED** [review]
Re-reviewed current head `1bfbbb2`. The latest push genuinely fixes the migration wiring/numbering, baseline-folded deposit indices, effective restore URL, full-tuple record rebuild, and hard errors returned by the outer restore calls. This is still **not merge-ready**:

1. **Baseline persistence is not failure-atomic.** `LetGateState::evaluate` mutates the in-memory baseline before `set_let_gate_baseline(...).await`. If that write fails, the process remains armed in memory and later ticks never retry persistence; restart can re-arm and absorb a standing gap.
2. **The claimed fail-closed scan can return `Ok(partial)`.** `recover_missed_bridge_outs` warns and continues on per-block `get_block_by_number` and `get_notes_by_id` failures, then returns a partial identity map. The outer `?` only catches total-call errors. Restore can therefore emit/skip incomplete history and then advance the projector cursor to the tip.
3. **Cross-transaction order is transaction-hash order, not execution order.** Node `sync_transactions` returns same-block rows ordered by `transaction_id`; this code assigns `tx_order` by response enumeration. For multiple bridge transactions in one block, authoritative order must be reconstructed by chaining each account update from `initial_state_commitment` to `final_state_commitment`, as miden-client does. Otherwise deposit indices/global indices can be swapped.
4. **Restore still cannot preserve same-details multiplicity.** Production miden-client SQLite keys input notes by `details_commitment`, so distinct NoteIds with the same details collapse before `get_input_notes(Consumed)`. The later `details_commitment -> Vec<NoteId>` queue cannot recreate a missing record; in-memory queue tests do not exercise this store boundary.

These are core Cantina #7 invariants, not documentation polish. Restore needs an authoritative NoteId-keyed body/identity stream end to end, execution-order reconstruction from account-state commitments, a real SQLite multiplicity regression, and persistence-failure barriers. CI is green, but `e2e-l2l2` is skipped.

File: docs/operations/let-cardinality-gate.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+missing note) resumes projection automatically at the exact blocks.
+
+## Diagnosing a halt
+
+1. Read the standing error: it carries `kind`, `gap`, `on_chain`, and the pending
+   visible count.
+2. Compare the ledgers yourself:
+   - on-chain: `read_let_num_leaves` (the #9 monitor logs it as `on_chain`);
+   - local: `SELECT COUNT(*) FROM synthetic_logs WHERE topics[1] = <BridgeEvent topic>`
+     (== `deposit_counter`) **plus** `SELECT COUNT(*) FROM unbridgeable_bridge_outs`.
```
[the remark this answers is not in the document]
Comment (mandrigin) [reply docs/operations/let-cardinality-gate.md]: Addressed in c431676: the diagnostic no longer derives LET accounting from synthetic-log plus quarantine counts. Durable reservations, including non-emitted leaves, are the accounting source.

File: docs/operations/let-cardinality-gate.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+     dropped BridgeEvent AND an unaccounted leaf);
+   - node/store rollback skew after a restore.
+4. `local_ahead` means the local store double-counted (crash-replay bug, foreign
+   note misattributed as bridge-consumed) — treat as corruption; do not restart
+   into it repeatedly, snapshot the store and escalate.
+
+## Relationship to the Cantina #9 monitor
+
+`run_let_divergence_check` (alarm-only, post-emit) remains as the independent
+second view — now quarantine-aware (`deposit_count + unbridgeable rows`). It still
```
[the remark this answers is not in the document]
Comment (mandrigin) [reply docs/operations/let-cardinality-gate.md]: Addressed in c431676: the old quarantine-aware divergence monitor and double-counting guidance were removed; quarantined leaves are already represented by their reservations.

File: src/synthetic_projector.rs [hunk 3]
```diff
@@ -1586,23 +1764,154 @@ impl SyntheticProjector {
...
+                        (None, Some(b)) => Some(b),
+                        _ => None,
+                    };
+                    (v, newly_armed, was_diverged)
+                };
+                // Persist the arming exactly once — a restart must re-evaluate against THIS
+                // baseline, never re-absorb a standing gap into a fresh one.
+                if let Some(baseline) = newly_armed {
+                    self.store.set_let_gate_baseline(baseline).await?;
+                    tracing::info!(
```
[the remark this answers is not in the document]
Comment (mandrigin) [reply src/synthetic_projector.rs]: Addressed in c431676: the mutable retry/arming gate was removed. The baseline is explicit operator-audited durable state, and each tick recomputes cardinality from persisted values before sealing.

File: src/restore.rs [hunk 4]
```diff
@@ -584,7 +705,85 @@ async fn recover_missed_bridge_outs(
...
+    let mut per_block_order: std::collections::HashMap<u64, u32> = std::collections::HashMap::new();
+    for tx in &txs {
+        if tx.transaction_header.account_id() != bridge_id {
+            continue;
+        }
+        let block = tx.block_num.as_u64();
+        let order = *per_block_order
+            .entry(block)
+            .and_modify(|i| *i += 1)
+            .or_insert(0u32);
```
[the remark this answers is not in the document]
Comment (mandrigin) [reply src/restore.rs]: Addressed in c431676: restore uses the same `ordered_account_transactions` execution-chain helper and preserves the authoritative input position for each replayed NoteId.

File: src/store/mod.rs [hunk 5]
```diff
@@ -641,6 +666,12 @@ pub trait Store: Send + Sync + 'static {
         Ok(true)
     }
 
+    /// Number of quarantined (unbridgeable) B2AGG rows — each one occupies an on-chain
+    /// LET leaf that deliberately has NO emitted BridgeEvent, so honest LET accounting is
+    /// `deposit_count + this` (Cantina #7/#9). Kept a COUNT (not a list fetch): it is read
+    /// on every divergence-monitor pass.
```
[the remark this answers is not in the document]
Comment (mandrigin) [reply src/store/mod.rs]: Addressed in c431676: the obsolete quarantine-count accounting API was removed. LET cardinality uses the reservation counter, which already includes non-emitted leaves.

File: src/synthetic_projector.rs [hunk 6]
```diff
@@ -1586,23 +1764,154 @@ impl SyntheticProjector {
...
+                // feed-visible B2AGG records (crash-replayed blocks are already in the
+                // counter via their reservations — checking each record's reservation keeps
+                // the recount from double-counting them).
+                let reserved_total = self.store.get_deposit_count().await?;
+                let mut pending_unreserved: u64 = 0;
+                for (id, _) in &auth_b2agg {
+                    if self
+                        .store
+                        .get_reserved_deposit_index(&id.to_hex())
+                        .await?
```
[the remark this answers is not in the document]
Comment (mandrigin) [reply src/synthetic_projector.rs]: Addressed in c431676: reservation lookup is now the batch `get_deposit_indices` store API, eliminating the per-note PostgreSQL N+1 reads.

File: src/synthetic_projector.rs [hunk 7]
```diff
@@ -1796,7 +2112,7 @@ pub(crate) fn bridge_consumed_nullifiers(
             .entry(block)
             .and_modify(|i| *i += 1)
             .or_insert(0u32);
```
[the remark this answers is not in the document]
Comment (mandrigin) [reply src/synthetic_projector.rs]: Addressed in c431676: `ordered_account_transactions` reconstructs same-block execution order by chaining initial/final account-state commitments and fails closed on an ambiguous or disconnected chain.

**Review by mandrigin — COMMENTED** [review]
Re-reviewed current head `c431676`; this supersedes my earlier blocking reviews. I found no remaining production correctness blocker in the scoped single-projector design.

Confirmed:
- same-block transactions are ordered by the account-state commitment chain, and same-transaction B2AGGs use the committed `TransactionRecord.input_notes()` sequence;
- unique NoteId identity survives same-details siblings and restart via the append-only nullifier ledger plus the one-time migration-017 backfill;
- every bridge-consumed leaf reserves its absolute LET index before non-emission gates, existing reservations must form the exact execution-order prefix, and cardinality is checked before sealing;
- incomplete identity/body responses fail before cursor advancement or sealing, including the already-reserved retry case;
- restore uses the authoritative NoteId/body/transaction stream with exact order and cardinality checks.

Validation: `make lint` passed; `make test-unit` passed (434 passed, 1 intentionally ignored); focused projector 33/33 and restore 35/35; Postgres-feature all-target check passed. E2E was intentionally left for the follow-up release run. Distributed PostgreSQL replicas remain out of scope.

**Review by IvanBelyakoff — DISMISSED** [review]

**Review by IvanBelyakoff — APPROVED** [review]
